qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -0,0 +1,322 @@
+            qc.append(psi_y_dag.to_instruction().inverse(), qc.qubits)
+
+            if measurement:
+                qc.barrier(q)
+                qc.measure(q, c)
+        return qc
+
+    def _compute_overlap(self, idx, results, is_statevector_sim, measurement_basis):
+        """
+        Helper function to compute overlap for given input.
+        """
+        if is_statevector_sim:
+            # |<0|Psi^dagger(y) x Psi(x)|0>|^2, take the amplitude
+            v_a, v_b = [results.get_statevector(int(i)) for i in idx]
+            tmp = np.vdot(v_a, v_b)
+            kernel_value = np.vdot(tmp, tmp).real  # pylint: disable=no-member
+        else:
+            result = results.get_counts(idx)
+
+            kernel_value = result.get(measurement_basis, 0) / sum(result.values())
+        return kernel_value
+
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        r"""
+        Construct kernel matrix for given data and feature map
+
+        If y is None, self inner product is calculated.
+        If using `statevector_simulator`, only build circuits for :math:`\Psi(x)|0\rangle`,
+        then perform inner product classically.
+
+        Args:
+            x_vec: 1D or 2D array of datapoints, NxD, where N is the number of datapoints,
+                                                            D is the feature dimension
+            y_vec: 1D or 2D array of datapoints, MxD, where M is the number of datapoints,
+                                                            D is the feature dimension
+
+        Returns:
+            2D matrix, NxM
+
+        Raises:
+            AlgorithmError:
+                - A quantum instance or backend has not been provided to the class
+            ValueError:
+                - x_vec and/or y_vec are not one or two dimensional arrays
+                - x_vec and/or y_vec have incompatible dimension with feature map
+        """
+        if self._quantum_instance is None:
+            raise AlgorithmError("A QuantumInstance or Backend "
+                            "must be supplied to run the quantum kernel.")
+        if isinstance(self._quantum_instance, (BaseBackend, Backend)):
+            self._quantum_instance = QuantumInstance(self._quantum_instance)
+
+        if not isinstance(x_vec, np.ndarray):
+            x_vec = np.asarray(x_vec)
+        if y_vec is not None and not isinstance(y_vec, np.ndarray):
+            y_vec = np.asarray(y_vec)
+
+        if x_vec.ndim > 2:
+            raise ValueError("x_vec must be a 1D or 2D array")
+
+        if x_vec.ndim == 1:
+            x_vec = np.reshape(x_vec, (-1, 2))
+
+        if y_vec is not None and y_vec.ndim > 2:
+            raise ValueError("y_vec must be a 1D or 2D array")
+
+        if y_vec is not None and y_vec.ndim == 1:
+            y_vec = np.reshape(y_vec, (-1, 2))
+
+        # TODO: update number of qubits --> should allow setting the number of params too!
```

Comment:
Will you be addressing this TODO?

Reply:
To be discussed with @stefan-woerner who added the TODO.

Reply:
This implicitly assumes that the number of qubits equals the number of parameters (as in the ZZFeatureMap). However, that doesn't always need to be the case. Here we should adjust as follows:
1. If `self._feature_map.num_parameters == x_vec.shape[1]` nothing will be done (cf. check in line 210).
2. if 1. does not hold, then we can check wether the feature map actually allows to set the `num_qubits` or not. If not, raise Exception, if yes, adjust and assume it sets the number of parameters correctly, if htis is not true, raise an exception.

Reply:
Turns out this assignment isn't required in the current Qiskit, (`self._feature_map.num_qubits = x_vec.shape[1]`) and subsequent checks for `self._feature_map.num_parameters` and `x_vec.shape[1]`, `y_vec.shape[1]` are sufficient.